### PR TITLE
Improved Retry and Sub Resource Support

### PR DESCRIPTION
**Pull request by csviri**
- retry funtionality
- explicit API for status sub-resource handling
- removed support for generatio awareness

### CODE REVIEW

File: operator-framework/src/main/java/com/github/containersolutions/operator/api/UpdateControl.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.util.Optional;
+
+public class UpdateControl<T extends CustomResource> {
+
+    private final T customResource;
+    private final boolean updateStatusSubResource;
+    private final boolean updateCustomResource;
+
+    private UpdateControl(T customResource, boolean updateStatusSubResource, boolean updateCustomResource) {
```
Comment (adam-sandor) [inline operator-framework/src/main/java/com/github/containersolutions/operator/api/UpdateControl.java:13]: very nicely done with the static methods

File: operator-framework/src/main/java/com/github/containersolutions/operator/processing/EventDispatcher.java [hunk 2]
```diff
@@ -100,14 +116,21 @@ private boolean markedForDeletion(CustomResource resource) {
     }
 
     // created to support unit testing
```
Comment (sjmittal) [inline operator-framework/src/main/java/com/github/containersolutions/operator/processing/EventDispatcher.java:118]: This comment seems bit misleading as this class in invoked from regular flow too.

File: operator-framework/src/main/java/com/github/containersolutions/operator/processing/EventDispatcher.java [hunk 3]
```diff
@@ -75,14 +91,14 @@ private boolean hasDefaultFinalizer(CustomResource resource) {
     }
 
     private void removeDefaultFinalizer(CustomResource resource) {
+        log.debug("Removing finalizer on {}: {}", resource);
```
Comment (sjmittal) [inline operator-framework/src/main/java/com/github/containersolutions/operator/processing/EventDispatcher.java:94]: If we log entire resource it becomes very verbose. Is it possible to use `trace` whenever we want to log entire resource.

File: operator-framework/src/main/java/com/github/containersolutions/operator/processing/EventDispatcher.java (hunk 2 above)
Comment (csviri) [reply operator-framework/src/main/java/com/github/containersolutions/operator/processing/EventDispatcher.java:113]: > This comment seems bit misleading as this class in invoked from regular flow too.

But its still true, and this is the only reason we are created this class. This is intentional.

File: operator-framework/src/main/java/com/github/containersolutions/operator/processing/EventDispatcher.java (hunk 3 above)
Comment (csviri) [reply operator-framework/src/main/java/com/github/containersolutions/operator/processing/EventDispatcher.java:87]: > If we log entire resource it becomes very verbose. Is it possible to use `trace` whenever we want to log entire resource.

You are right, could you pls create an issue to sanitize logs?

**Comment by csviri** [comment]
https://github.com/ContainerSolutions/java-operator-sdk/issues/136
this would be fixed also by this PR

**Comment by adam-sandor** [comment]
integration tests are failing when I run them agains a GKE cluster

**Pull request merged**